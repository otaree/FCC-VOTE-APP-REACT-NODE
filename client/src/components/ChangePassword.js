import React from 'react';
import { connect } from 'react-redux';
import axios from 'axios';

import * as authActions from '../store/actions/auth';


export class ChangePassword extends React.Component {
    state = {
        password: {
            value: "",
            error: false
        },
        password2: {
            value: "",
            error: false
        }
    };

    passwordChangeHandler = e => {
        this.setState({ password: {
            value: e.target.value,
            error: false
        }});
    };

    password2ChangeHandler = e => {
        this.setState({ password2: {
            value: e.target.value,
            error: false
        }});
    };

    submitHandler = async e => {
        e.preventDefault();

        if (this.validation()) return;

        try {
            await axios({ method: "patch", url: "http://localhost:5000/user/password/change", data: { password: this.state.password.value }, headers: { 'x-auth': this.props.token } });
            this.props.history.push('/');
        } catch (e) {
            this.props.changeFail();
        }
    };

    validation = () => {
        let error = false;

        if (this.state.password.value.length < 6) {
            error = true;
            this.setState(prevState => {
                return {
                    password: {
                        value: prevState.password.value,
                        error: true
                    }
                };
            });
        }

        if (this.state.password.value !== this.state.password2.value) {
            error = true;
            this.setState(prevState => {
                return {
                    password2: {
                        value: prevState.password2.value,
                        error: true
                    }
                };
            });
        }

        return error;
    };

    render() {
        return (
            <section className="auth">
                {
                    this.props.authError.isError &&
                    <div className="auth-server-fail">{this.props.authError.value}</div>
                }
                <form className="form" onSubmit={this.submitHandler}>
                    <div className="form-control">
                        <input 
                            type="password"
                            placeholder="New Password"
                            value={this.state.password.value} 
                            onChange={this.passwordChangeHandler}
                        />
                        {
                             this.state.password.error &&
                             <div className="auth-error">Invalid password</div>
                         }
                    </div>
                    <div className="form-control">
                        <input 
                            type="password"
                            placeholder="Confirm Password"   
                            value={this.state.password2.value} 
                            onChange={this.password2ChangeHandler}
                        />
                        {
                             this.state.password2.error &&
                             <div className="auth-error">Passwords dose not match</div>
                         }
                    </div>
                    <div className="form-control">
                        <input 
                            className="btn"
                            type="submit"
                            value="Change Password"    
                        />
                    </div>
                </form>
            </section>
        );
    }
}

const mapStateToProps = state => {
    return {
        authError: state.error,
        loading: state.loading,
        token: state.token
    };
};

const mapDispatchToProps = dispatch => {
    return {
        changeFail: () => dispatch(authActions.fail("Network Problem"))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ChangePassword);