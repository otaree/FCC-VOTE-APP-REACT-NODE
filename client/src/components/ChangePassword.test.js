import React from 'react';
import { shallow } from 'enzyme';

import { ChangePassword } from './ChangePassword';

let props, changePassword;
describe('ChangePassword', () => {
    props = {
        authError: {
            isError: false,
            value: ''
        }
    };
    changePassword = shallow(<ChangePassword {...props} />);

    it('renders correctly', () => {
        expect(changePassword.find('.auth-server-fail').length).toEqual(0); 
        expect(changePassword).toMatchSnapshot();
    });

    describe('When from fields value changes', () => {
        beforeEach(() => {
            changePassword = shallow(<ChangePassword {...props} />);
        });

        it('should set password on input change', () => {
            const value = "password!";
            changePassword.find('input').at(0).simulate('change', {
                target: { value }
            });
            expect(changePassword.state('password').value).toEqual(value);
        });

        it('should set password2 on input change', () => {
            const value = "password!";
            changePassword.find('input').at(1).simulate('change', {
                target: { value }
            });
            expect(changePassword.state('password2').value).toEqual(value);
        });
    });
});