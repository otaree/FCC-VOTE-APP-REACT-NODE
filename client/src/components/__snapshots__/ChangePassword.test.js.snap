// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePassword renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChangePassword
    authError={
        Object {
            "isError": false,
            "value": "",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <form
          className="form"
          onSubmit={[Function]}
>
          <div
                    className="form-control"
          >
                    <input
                              onChange={[Function]}
                              placeholder="New Password"
                              type="password"
                              value=""
                    />
          </div>
          <div
                    className="form-control"
          >
                    <input
                              onChange={[Function]}
                              placeholder="Confirm Password"
                              type="password"
                              value=""
                    />
          </div>
          <div
                    className="form-control"
          >
                    <input
                              className="btn"
                              type="submit"
                              value="Change Password"
                    />
          </div>
</form>,
      ],
      "className": "auth",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-control"
>
              <input
                            onChange={[Function]}
                            placeholder="New Password"
                            type="password"
                            value=""
              />
</div>,
            <div
              className="form-control"
>
              <input
                            onChange={[Function]}
                            placeholder="Confirm Password"
                            type="password"
                            value=""
              />
</div>,
            <div
              className="form-control"
>
              <input
                            className="btn"
                            type="submit"
                            value="Change Password"
              />
</div>,
          ],
          "className": "form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="New Password"
                  type="password"
                  value=""
/>,
                false,
              ],
              "className": "form-control",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "New Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Confirm Password"
                  type="password"
                  value=""
/>,
                false,
              ],
              "className": "form-control",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Confirm Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="btn"
                type="submit"
                value="Change Password"
/>,
              "className": "form-control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "btn",
                "type": "submit",
                "value": "Change Password",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <form
            className="form"
            onSubmit={[Function]}
>
            <div
                        className="form-control"
            >
                        <input
                                    onChange={[Function]}
                                    placeholder="New Password"
                                    type="password"
                                    value=""
                        />
            </div>
            <div
                        className="form-control"
            >
                        <input
                                    onChange={[Function]}
                                    placeholder="Confirm Password"
                                    type="password"
                                    value=""
                        />
            </div>
            <div
                        className="form-control"
            >
                        <input
                                    className="btn"
                                    type="submit"
                                    value="Change Password"
                        />
            </div>
</form>,
        ],
        "className": "auth",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form-control"
>
                <input
                                onChange={[Function]}
                                placeholder="New Password"
                                type="password"
                                value=""
                />
</div>,
              <div
                className="form-control"
>
                <input
                                onChange={[Function]}
                                placeholder="Confirm Password"
                                type="password"
                                value=""
                />
</div>,
              <div
                className="form-control"
>
                <input
                                className="btn"
                                type="submit"
                                value="Change Password"
                />
</div>,
            ],
            "className": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="New Password"
                    type="password"
                    value=""
/>,
                  false,
                ],
                "className": "form-control",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "New Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Confirm Password"
                    type="password"
                    value=""
/>,
                  false,
                ],
                "className": "form-control",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Confirm Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="btn"
                  type="submit"
                  value="Change Password"
/>,
                "className": "form-control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "btn",
                  "type": "submit",
                  "value": "Change Password",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
